WIP: Add WithDockerEndpoint overload with credentials

prskr opened the pull request:
Please note that this is a work in progress PR to get early feedback.

I'll squash all commits according to the commit message format described in the guidelines as soon as the PR is completed and ready to be merged.

prskr committed:
WIP introduce WithDockerEndpoint overload with credentials

prskr committed:
Prepare cert auth setup based on env variables

prskr committed:
Cert loading showcase

HofmeisterAn commented on global.json:
```diff
@@ -1,6 +1,6 @@
 {
```
Do not commit this change or update the pipeline as well.

prskr replied:
Updated the Azure pipeline config - anything else?

HofmeisterAn replied:
Yes, your change in `azure-pipelines.yml` is correct. But it would be enough to keep `5.0.200` in `global.json`. See [rollForward](https://docs.microsoft.com/de-de/dotnet/core/tools/global-json?tabs=netcore3x#rollforward) (latestPatch).

prskr replied:
Interesting, had some issues when loading the project. I'll try to revert the change here but I'd keep the latest version in the Azure pipeline, no reason to keep the old one I guess?

HofmeisterAn replied:
>Interesting, had some issues when loading the project

That's on purpose if you do not have a compatible patch Version. This will prevent different behaviors with different Versions of .NET Core. This file pins the exact version.

>  I'll try to revert the change here

I have added some changes to your forked branch by pull request. I included that one as well.

> no reason to keep the old one I guess?

Yep, we can update the pipeline to a newer version.

HofmeisterAn commented on src/DotNet.Testcontainers.Tests/Unit/Containers/TestcontainersAccessInformationTest.cs:
```diff
@@ -2,6 +2,7 @@ namespace DotNet.Testcontainers.Tests.Unit.Containers
 {
   using System;
   using System.Threading.Tasks;
+  using Docker.DotNet;
```
`using Docker.DotNet;` is not necessary.

HofmeisterAn commented on src/DotNet.Testcontainers/Clients/DockerApiEndpoint.cs:
```diff
@@ -1,10 +1,54 @@
+  using System.Linq;
+  using Docker.DotNet;
+  using Docker.DotNet.X509;
+  using Services;
```
Use full namespaces.

HofmeisterAn commented on src/DotNet.Testcontainers/Clients/DockerApiEndpoint.cs:
```diff
@@ -1,10 +1,54 @@
+
+    public static Uri Default => TestcontainersHostService.GetService<IOperatingSystem>().DockerApiEndpoint;
+
+    public static Credentials DefaultCredentials => DetermineCredentialsFromEnv();
```
~~This should be part of `TestcontainersBuilder`. The first task I mentioned simply allows the configuration for the client certificate authentication through the builder pattern. Like it's done for every other case~~. I have not expressed myself very well here. IMO it should be more like `IOperatingSystem` and part of its own class.

HofmeisterAn commented on src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs:
```diff
@@ -6,18 +6,19 @@ namespace DotNet.Testcontainers.Clients
+  using Containers.Configurations;
+  using Containers.OutputConsumers;
+  using Internals.Mappers;
+  using Services;
```
Use full namespaces.

HofmeisterAn commented on src/DotNet.Testcontainers/Clients/DockerImageOperations.cs:
```diff
@@ -6,15 +6,16 @@ namespace DotNet.Testcontainers.Clients
+  using Containers.Configurations;
+  using Images;
+  using Images.Archives;
+  using Images.Configurations;
```
Use full namespaces.

HofmeisterAn commented on src/DotNet.Testcontainers/Clients/DockerSystemOperations.cs:
```diff
@@ -3,10 +3,12 @@ namespace DotNet.Testcontainers.Clients
   using System.Threading;
   using System.Threading.Tasks;
+  using Docker.DotNet;
+  using JetBrains.Annotations;
```
`using JetBrains.Annotations;` is not necessary.

HofmeisterAn commented on src/DotNet.Testcontainers/Containers/Builders/ITestcontainersBuilder.cs:
```diff
@@ -176,6 +177,16 @@ public interface ITestcontainersBuilder<out TDockerContainer>
+    /// <param name="credentials"></param>
+    /// <returns></returns>
+    [PublicAPI]
+    ITestcontainersBuilder<TDockerContainer> WithDockerEndpoint(string endpoint, Credentials credentials);
```
Do not use Docker remote API specific types here.  To keep the API more stable use primitive types or an own interface for .NET Testcontainers (like `WithRegistryAuthentication`, `IDockerImage`, etc.).

HofmeisterAn commented on src/DotNet.Testcontainers/Containers/Builders/TestcontainersBuilder.cs:
```diff
@@ -7,6 +7,7 @@ namespace DotNet.Testcontainers.Containers.Builders
   using System.Reflection;
   using System.Threading;
   using System.Threading.Tasks;
+  using Docker.DotNet;
```
Do not use Docker remote API specific types here.

HofmeisterAn commented on src/DotNet.Testcontainers/Containers/Builders/TestcontainersBuilder.cs:
```diff
@@ -238,7 +246,8 @@ public TDockerContainer Build()
       return new TestcontainersConfiguration(
-        endpoint ?? DockerApiEndpoint.Local,
+        endpoint ?? DockerApiEndpoint.Default,
+        endpointCredentials ?? DockerApiEndpoint.DefaultCredentials,
```
Replace this with:

```csharp
return new TestcontainersConfiguration(
  endpoint,
  endpointCredentials,
  ...,
```

and apply the default configuration at Line:44 (maybe similar to `AuthenticationConfiguration`?):

```csharp
public TestcontainersBuilder() : this(
  Apply(
    authConfig: new AuthenticationConfiguration(),
    labels: new DefaultLabels(),
    endpoint: .... /* Something like that. */
    useTlsVerification: ....
    certificatePath: ....
    outputConsumer: Consume.DoNotConsumeStdoutAndStderr(),
    waitStrategies: Wait.ForUnixContainer().Build(),
    startupCallback: (_, _) => Task.CompletedTask),
  _ => { })
{
}
```

If one of the env variables are set, apply the configuration here.

HofmeisterAn commented on src/DotNet.Testcontainers/Containers/Configurations/ITestcontainersConfiguration.cs:
```diff
@@ -4,6 +4,7 @@ namespace DotNet.Testcontainers.Containers.Configurations
   using System.Collections.Generic;
   using System.Threading;
   using System.Threading.Tasks;
+  using Docker.DotNet;
```
Do not use Docker remote API specific types here.

HofmeisterAn commented on src/DotNet.Testcontainers/Containers/Configurations/TestcontainersConfiguration.cs:
```diff
@@ -4,6 +4,7 @@ namespace DotNet.Testcontainers.Containers.Configurations
   using System.Collections.Generic;
   using System.Threading;
   using System.Threading.Tasks;
+  using Docker.DotNet;
```
Do not use Docker remote API specific types here.

HofmeisterAn commented on src/DotNet.Testcontainers/DotNet.Testcontainers.csproj:
```diff
@@ -4,6 +4,7 @@
     <LangVersion>latest</LangVersion>
   </PropertyGroup>
   <ItemGroup>
+    <PackageReference Include="Docker.DotNet.X509" Version="3.125.2" />
```
Put it below `Docker.DotNet`.

HofmeisterAn commented on src/DotNet.Testcontainers/Services/IOperatingSystem.cs:
```diff
@@ -7,6 +7,8 @@ namespace DotNet.Testcontainers.Services
   /// </summary>
   internal interface IOperatingSystem
   {
+    internal const string DockerHostEnvName = "DOCKER_HOST";
```
Register it as an implementation of IOperatingSystem in TestcontainersHostService. Resolve it first if the env variable is set. Maybe we can use the same approach for the other env variables too.

Something like (maybe we can improve this)? What do you think?

```csharp
internal readonly struct DockerHostOperatingSystem : IOperatingSystem
{
  private readonly IOperatingSystem os;

  public DockerHostOperatingSystem(IOperatingSystem os)
  {
    this.os = os;
    this.DockerApiEndpoint = new Uri(Environment.GetEnvironmentVariable("DOCKER_HOST") ?? this.os.DockerApiEndpoint.AbsoluteUri);
  }

  public Uri DockerApiEndpoint { get; }

  public string NormalizePath(string path)
  {
    return this.os.NormalizePath(path);
  }
}
```

HofmeisterAn commented on src/DotNet.Testcontainers/Services/Unix.cs:
```diff
@@ -1,13 +1,16 @@
   internal readonly struct Unix : IOperatingSystem
   {
-    private static readonly Uri endpoint = new Uri("unix:/var/run/docker.sock");
+    private const string DefaultDockerSocketPath = "unix:/var/run/docker.sock";
```
See `IOperatingSystem`.

HofmeisterAn commented on src/DotNet.Testcontainers/Services/Windows.cs:
```diff
@@ -9,7 +10,8 @@ namespace DotNet.Testcontainers.Services
 #pragma warning disable S1075
 
-    private static readonly Uri endpoint = new Uri("npipe://./pipe/docker_engine");
+    private const string DefaultDockerSocketPath = "npipe://./pipe/docker_engine";
```
See `IOperatingSystem`.

HofmeisterAn reviewed (changes requested):
Not that much to change. Just a few requests to keep the implementation more consistent:

- Do not use `Docker.DotNet` in the mentioned classes. I like to hide it, to keep the API more stable.
- Move the default configuration into `TestcontainersBuilder` line 44.
    - Maybe put the certificate configuration into its own class and register it. Similar to `IOperatingSystem`.
    I attached an example for `DOCKER_HOST` below.

Thanks for your work and the pull request.

prskr committed:
Apply some nits and refactor API to not use any Docker client related classes

prskr commented:
I updated the PR and tried to apply as much of your suggestions as possible.

Right now this would be a breaking change but I'm not sure how you would like to avoid this therefore I thought the easiest way to discuss this would be to push the already broken code and see if we can find a good solution if there is any.

I'm also not that satisfied with the naming of the `DockerClientAuthConfig` and I thought it me be even better to make a breaking change now and introduce something like a `DockerClientConfig` that also includes the endpoint URI to be able to extend it in the future.

The cert loading is still under construction because I'd like to actually parse the certificate labels to not introduce kinda 50% solution.

I hope I applied the `IOperatingSystem` suggestion correct.

HofmeisterAn commented:
> Right now this would be a breaking change but I'm not sure how you would like to avoid this therefore I thought the easiest way to discuss this would be to push the already broken code and see if we can find a good solution if there is any.

I'll take a look at it in the next days. I'm pretty sure we will find an implementation that won't be a breaking change.

> I'm also not that satisfied with the naming of the DockerClientAuthConfig and I thought it me be even better to make a breaking change now and introduce something like a DockerClientConfig that also includes the endpoint URI to be able to extend it in the future.

I try to add my suggestions on top of your pull request.

> The cert loading is still under construction because I'd like to actually parse the certificate labels to not introduce kinda 50% solution.

👍

HofmeisterAn committed:
Add IDockerClientAuthenticationConfiguration; Clean up project.

HofmeisterAn committed:
Remove unnecessary usings.

HofmeisterAn committed:
Add DockerClientEnvironmentAuthenticationConfiguration; Load default IDockerClientAuthenticationConfiguration; Add sample tests.

prskr committed:
Merge pull request #1 from HofmeisterAn/baez90-feature/client-cert-auth

Add IDockerClientAuthenticationConfiguration; Clean up project.

prskr committed:
Move to Credentials class

- add PemCertificateCredentials to load all kind of PEM certs
- add some very basic tests to ensure loading works
- move cert file discovery to DockerCertDir
- properly distinguish between TLS_VERIFY and client auth

HofmeisterAn commented on global.json:
```diff
@@ -1,6 +1,6 @@
 {
   "sdk": {
-    "version": "5.0.100",
+    "version": "5.0.103",
```
This should be `5.0.200` again 😀.

prskr replied:
I know :sweat_smile: I'll try to keep this in mind but I worked on a different machine yesterday with a different patch level and I wanted to see if I understood the config option in the `global.json` correctly (I did not :sweat_smile: ) therefore I kept it like this but eventually I will make sure it matches the one in the Azure pipeline

HofmeisterAn replied:
> I worked on a different machine yesterday with a different patch level and I wanted to see if I understood the config option in the `global.json` correctly (I did not 😅)

`200` is not the patch version. It's a bit different than semantic versioning. It's `x.y.znn`:

- `x` is the major version.
- `y` is the minor version.
- `z` is the feature band.
- `nn` is the patch version.

`5.0.100` is valid for the local installed sdks `5.0.101`, `5.0.102`, `5.0.103`, ... **but** not for `>= 5.0.200`.
I hope that helps 😅.

prskr replied:
Thanks for clarification!
Looking back it feels kinda stupid expecting Microsoft to stick to something like semantic versioning - of course there's more! :roll_eyes: but at least I do understand now why I broke stuff...

prskr commented:
I did some work yesterday to move the cert loading to `PemCertificateCredentials` and to move back to `Credentials`.
To be honest I'm not sure why I'd need another implementation of `IDockerClientConfiguration` because most of the logic required to support other credentials could be move either to constructors (e.g. one that takes username, password to initialize HTTP basic credentials) or to custom `Credentials` implementations like the `PemCertificateCredentials`.

I would now add a base class for `PemCertificateCredentials` and `X509CertificateCredentials` that takes `(bool tlsVerify, X509Certificate2 caCert)` to handle the certificate validation at the base class and then all the magic for handling different cert mechanisms goes to inherited classes? Or do you prefer a different approach?

BTW: the certs I added are just random generated ones from Docker-in-Docker for testing the cert loading mechanism.

HofmeisterAn reviewed (commented):
> I would now add a base class for `PemCertificateCredentials` and `X509CertificateCredentials` that takes `(bool tlsVerify, X509Certificate2 caCert)` to handle the certificate validation at the base class and then all the magic for handling different cert mechanisms goes to inherited classes? Or do you prefer a different approach?

Sounds good. I'll resolve the conversations from the previous review and look at your changes if your done with `PemCertificateCredentials` and `X509CertificateCredentials`, ok? I took a quick look at the changes from yesterday. Looks good so far 👍 . As I said, if you need any help -  let me know.

prskr committed:
Replace PemCertificateCredentials with TlsCredentials

Applied a more generic way to handle all different cases:

* TLS connection without credentials and CA certificate
* TLS with custom CA certificate but without credentials
* disable TLS verification without credentials
* TLS without custom CA certificate but with credentials
* TLS with custom CA and credentials

and this does not cover basic authentication at this point.

Moved PEM reading to a wrapper class that also takes care of indexing the certificates directory.

prskr commented:
Tests are still missing at this point but I thought it might be good to first check if the design is okay for you before putting too much effort into something that does not fit the overall architecture.

I actually ditched the whole x509 NuGet package and implemented the whole thing myself after having a look at the original implementation which isn't flexible enough to allow all different variants that are actually possible - or I completely missed something but I think the current solution fits at least okay-ish into the project, covers all cases and is still flexible enough to be extended later on if necessary.

I won't be able to continue on this before next week and if you haven't had time till then to look into it I'll continue to write some tests anyway to see if I missed something :see_no_evil: hence - no reason to hurry! :blush: 
Thanks for all your effort!

prskr commented on src/DotNet.Testcontainers/Containers/Configurations/DockerClientConstants.cs:
```diff
@@ -0,0 +1,9 @@
+namespace DotNet.Testcontainers.Containers.Configurations
+{
+  internal static class DockerClientConstants
```
TODO: check if these can be moved to `DockerCertificatesDirectory`

HofmeisterAn commented:
> Tests are still missing at this point but I thought it might be good to first check if the design is okay for you before putting too much effort into something that does not fit the overall architecture.
> 
> I actually ditched the whole x509 NuGet package and implemented the whole thing myself after having a look at the original implementation which isn't flexible enough to allow all different variants that are actually possible - or I completely missed something but I think the current solution fits at least okay-ish into the project, covers all cases and is still flexible enough to be extended later on if necessary.
> 
> I won't be able to continue on this before next week and if you haven't had time till then to look into it I'll continue to write some tests anyway to see if I missed something 🙈 hence - no reason to hurry! 😊
> Thanks for all your effort!

Sorry – I had a very busy week. I needed some break from coding. I will look at it next week during the holidays.

HofmeisterAn commented on .editorconfig:
```diff
@@ -196,6 +196,10 @@ csharp_prefer_static_local_function = true:warning
 csharp_prefer_simple_using_statement = true:suggestion
 dotnet_diagnostic.IDE0063.severity = suggestion
 
+resharper_csharp_space_in_singleline_method = true
+resharper_csharp_space_between_accessors_in_singleline_property = true
+resharper_csharp_space_within_single_line_array_initializer_braces = true
```
The end of `.editorconfig` contains a dedicated block for ReSharper / Rider:

```
##########################################
# JetBrains Rider
##########################################
```

HofmeisterAn commented on src/DotNet.Testcontainers.Tests/Unit/Containers/DockerCertDirTest.cs:
```diff
@@ -0,0 +1,23 @@
+namespace DotNet.Testcontainers.Tests.Unit.Containers
+{
+  using System.IO;
+  using Testcontainers.Containers.Configurations;
```
Full namespace.

HofmeisterAn commented on src/DotNet.Testcontainers.Tests/Unit/Containers/DockerClientConfigurationTest.cs:
```diff
@@ -0,0 +1,34 @@
+namespace DotNet.Testcontainers.Tests.Unit.Containers
+{
+  using System;
+  using Testcontainers.Containers.Configurations;
```
Full namespace. Haven't reviewed anything else, test class is still in progress.

HofmeisterAn commented on src/DotNet.Testcontainers.Tests/Unit/Containers/TlsCredentialsTest.cs:
```diff
@@ -0,0 +1,17 @@
+namespace DotNet.Testcontainers.Tests.Unit.Containers
+{
+  using System.IO;
+  using Testcontainers.Containers.Configurations;
```
Full namespace.

HofmeisterAn commented on src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs:
```diff
@@ -97,12 +97,7 @@ public async Task AttachAsync(string id, IOutputConsumer outputConsumer, Cancell
-        Stderr = true,
-        Stream = true,
-      };
+      var attachParameters = new ContainerAttachParameters { Stdout = true, Stderr = true, Stream = true };
```
Remove unnecessary changes (formatting).

HofmeisterAn commented on src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs:
```diff
@@ -121,8 +116,9 @@ public async Task<long> ExecAsync(string id, IList<string> command, Cancellation
 
-      for (ContainerExecInspectResponse response; (response = await this.Docker.Exec.InspectContainerExecAsync(created.ID, ct)
-        .ConfigureAwait(false)) != null;)
+      for (ContainerExecInspectResponse response;
```
Remove unnecessary changes (formatting).

HofmeisterAn commented on src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs:
```diff
@@ -139,9 +135,9 @@ public async Task<string> RunAsync(ITestcontainersConfiguration configuration, C
-        Mounts = converter.Mounts,
+        // TODO: Should we keep this? If the Docker daemon remove containers we're no longer able to call e. g. `CleanUp(true)` + `GetExitCode()`.
+        // AutoRemove = configuration.CleanUp,
+        PortBindings = converter.PortBindings, Mounts = converter.Mounts
```
```
PortBindings = converter.PortBindings,
Mounts = converter.Mounts
```

HofmeisterAn commented on src/DotNet.Testcontainers/Containers/Builders/TestcontainersBuilder.cs:
```diff
@@ -7,11 +7,13 @@ namespace DotNet.Testcontainers.Containers.Builders
   using DotNet.Testcontainers.Containers.OutputConsumers;
   using DotNet.Testcontainers.Containers.WaitStrategies;
   using DotNet.Testcontainers.Images;
+  using DockerClientConfiguration = Configurations.DockerClientConfiguration;
```
Remove unnecessary change.

HofmeisterAn commented on src/DotNet.Testcontainers/Containers/Builders/TestcontainersBuilder.cs:
```diff
@@ -170,20 +173,33 @@ public ITestcontainersBuilder<TDockerContainer> WithMount(string source, string
     {
-      return Build(this, Apply(cleanUp: cleanUp))
-        .WithLabel(TestcontainersClient.TestcontainersCleanUpLabel, cleanUp.ToString().ToLowerInvariant());
+      return Build(this, Apply(cleanUp: cleanUp)).WithLabel(TestcontainersClient.TestcontainersCleanUpLabel, cleanUp.ToString().ToLowerInvariant());
```
Remove unnecessary changes (formatting).

HofmeisterAn commented on src/DotNet.Testcontainers/Containers/Builders/TestcontainersBuilder.cs:
```diff
@@ -306,7 +322,7 @@ public TDockerContainer Build()
 
     /// <summary>
-    /// Returns the changed Testcontainer configuration object. If there is no change, the previous Testcontainer configuration object is returned.
+    ///   Returns the changed Testcontainer configuration object. If there is no change, the previous Testcontainer configuration object is returned.
```
Remove unnecessary changes (formatting).

HofmeisterAn commented on src/DotNet.Testcontainers/Containers/Builders/TestcontainersBuilder.cs:
```diff
@@ -320,7 +336,7 @@ private static T Merge<T>(T next, T previous, T defaultConfiguration = null)
 
     /// <summary>
-    /// Merges all existing and new Testcontainer configuration changes. If there are no changes, the previous Testcontainer configurations are returned.
+    ///   Merges all existing and new Testcontainer configuration changes. If there are no changes, the previous Testcontainer configurations are returned.
```
Remove unnecessary changes (formatting).

HofmeisterAn commented on src/DotNet.Testcontainers/Containers/Builders/TestcontainersBuilder.cs:
```diff
@@ -333,14 +349,12 @@ private static IEnumerable<T> Merge<T>(IEnumerable<T> next, IEnumerable<T> previ
 
     /// <summary>
-    /// Merges all existing and new Testcontainer configuration changes. If there are no changes, the previous Testcontainer configurations are returned.
+    ///   Merges all existing and new Testcontainer configuration changes. If there are no changes, the previous Testcontainer configurations are returned.
```
Remove unnecessary changes (formatting).

HofmeisterAn commented on src/DotNet.Testcontainers/Containers/Configurations/DockerClientEnvironmentConfiguration.cs:
```diff
@@ -0,0 +1,78 @@
+    {
+      this.Endpoint = clientEndpoint;
+      var dockerCertificates = new DockerCertificatesDirectory(certificatesDirectory);
+      this.Credentials = new TlsCredentials(dockerCertificates.CaCertificate, dockerCertificates.ClientCertificate, isTlsVerificationEnabled, isTlsVerificationEnabled);
```
Haven't tested it yet, but does it work if e. g. only `DOCKER_HOST` is set?

HofmeisterAn commented on src/DotNet.Testcontainers/Containers/Builders/TestcontainersBuilder.cs:
```diff
@@ -170,20 +173,33 @@ public ITestcontainersBuilder<TDockerContainer> WithMount(string source, string
+    public ITestcontainersBuilder<TDockerContainer> WithDockerEndpoint(string clientEndpoint, bool tlsVerify = true)
     {
-      return Build(this, Apply(endpoint: new Uri(endpoint)));
+      var endpoint = new Uri(clientEndpoint);
```
Maybe put both codes to create `clientAuthConfig` into a static method.

HofmeisterAn reviewed (changes requested):
> Tests are still missing at this point but I thought it might be good to first check if the design is okay for you before putting too much effort into something that does not fit the overall architecture.

I think the overall implementation looks good! This allows different implementations of `IDockerClientConfiguration` in the future too. Just a few code formatting and consistency issues are left (I stopped at some point 😀). I can take care of them as soon as the implementation is ready.

**Edit**: Maybe it's better to split the pull request into at least two parts. In the first pull request we prepare the `IDockerClientConfiguration` and update the current implementation to use the interface. In the next pull requests we implement the different `Credentials`.

HofmeisterAn commented:
I am closing this PR due to inactivity.

prskr commented:
Yeah the use case I have/had got postponed hence I prioritized other things over this.
Is this still of interest or is there no other one requiring this? :sweat_smile:

HofmeisterAn commented:
> Is this still of interest or is there no other one requiring this? 😅

I think it's still necessary, but as long as no one is working on this feature, it doesn't need to be open. It's probably better to slice it into smaller chunks, too. First part could be just the interface to pass the necessary properties to the client, etc.
